File: mobx/src/main/java/com/ivansadovyi/mobx/Reaction.java
package com.ivansadovyi.mobx;

public class Reaction<R> implements Disposable {

	private Disposable disposable;
	private boolean wasFirstAutorunOccurred = false;

	public Reaction(final ReactionDataCallback<R> dataCallback, final ReactionSideEffectCallback<R> sideEffectCallback) {
		disposable = new Autorun(new Runnable() {
			@Override
			public void run() {
				R data = dataCallback.getData();

				if (wasFirstAutorunOccurred) {
					sideEffectCallback.onReaction(data);
				}

				wasFirstAutorunOccurred = true;
			}
		});
	}

	@Override
	public void dispose() {
		disposable.dispose();
	}
}

File: mobx/src/main/java/com/ivansadovyi/mobx/Reaction.kt
package com.ivansadovyi.mobx

typealias ReactionDataCallback<T> = () -> T
typealias ReactionSideEffectCallback<T> = (T) -> Unit

class Reaction<T>(
	dataCallback: ReactionDataCallback<T>,
	sideEffectCallback: ReactionSideEffectCallback<T>,
	fireImmediately: Boolean = false
) : Disposable {

	private val autorunDisposable: Disposable
	private var wasFired = false
	private var previousData: T? = null

	init {
		this.autorunDisposable = autorun {
			val data = dataCallback()
			val wasDataChanged = data != previousData

			if (!wasFired && fireImmediately) {
				sideEffectCallback(data)
			} else if (wasFired && wasDataChanged) {
				sideEffectCallback(data)
			}

			wasFired = true
			previousData = data
		}
	}

	override fun dispose() {
		this.autorunDisposable.dispose()
	}
}
